Save runtimes in configmaps

rakesh-garimella opened the pull request:
Work in Progress

ngtuna committed:
fix topic missing in pubsub tests

ngtuna committed:
separate topics for python versions

rakesh-garimella committed:
Working prototype

rakesh-garimella committed:
Added kubeless-config yaml

rakesh-garimella committed:
Cleanup

rakesh-garimella committed:
Working after cleanup

rakesh-garimella committed:
Modify the client to read the configmap

 - The configmap needs to be read by the client

andresmgot commented on pkg/controller/controller.go:
```diff
@@ -88,6 +89,9 @@ func New(cfg Config, smclient *monitoringv1alpha1.MonitoringV1alpha1Client) *Con
 	})
 
+	var l langruntime.Langruntimes
+	l.ReadConfigMap(cfg.KubeCli)
```
`ReadConfigMap` is too generic when it is retrieving the runtime info. Also the function should retrieve more info like the namespace of the configmap and the name. My suggestion is to rename it as the method `New` of the `Langruntimes`:

```
var l = langruntimes.New(cfg.KubeCli, 'kubeless', 'kubeless-config')
```

andresmgot commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,83 +7,71 @@ import (
+)
 
-var pythonVersions, nodeVersions, rubyVersions, dotnetcoreVersions []runtimeVersion
 var availableRuntimes []RuntimeInfo
```
`availableRuntimes` should be a property of the Langruntimes struct

andresmgot commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,83 +7,71 @@ import (
+
+// ImageSecrets for pulling the image
+type ImageSecrets struct {
+	ImageSecret string `yaml:"imageSecret,omitempty"`
```
better call it 
```
ImageSecret string `yaml:"imageSecret,omitempty"` 
```
so it has a clear name for its purpose

andresmgot commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,83 +7,71 @@ import (
+}
+
 // GetRuntimes returns the list of available runtimes as strings
 func GetRuntimes() []string {
```
all the methods should now be part of the `Langruntime` struct:
```
func (l *Langruntime) GetRuntimes() []string{
 	result := []string{}
  	for _, runtimeInf := range l.availableRuntimes {
...
````

andresmgot commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,83 +7,71 @@ import (
 }
 
+type runtimeVersion struct {
+	Name             string         `yaml:"name"`
```
try to avoid to export properties if they are not going to be accessible from the outside

andresmgot commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,83 +7,71 @@ import (
+}
+
+// ImageSecrets for pulling the image
+type ImageSecrets struct {
```
the same for the types, do not export them if they are not going to be used in a different package

andresmgot commented on pkg/utils/k8sutil.go:
```diff
@@ -684,6 +684,19 @@ func EnsureFuncDeployment(client kubernetes.Interface, funcObj *kubelessApi.Func
+
+	// Add the imagesecrets if present to pull images from private docker registry
+	imageSecrets, err := langruntime.GetImageSecrets(funcObj.Spec.Runtime)
+
```
you missed the `if err {...` here

andresmgot commented on pkg/utils/k8sutil.go:
```diff
@@ -684,6 +684,19 @@ func EnsureFuncDeployment(client kubernetes.Interface, funcObj *kubelessApi.Func
+	if len(imageSecrets) > 0 {
+		var lors []v1.LocalObjectReference
+		for _, s := range imageSecrets {
+			lor := v1.LocalObjectReference{Name: s}
```
It is better for `GetImageSecrets` to return a `[]v1.LocalObjectReference` rather than an array of strings.

andresmgot reviewed (commented):
This is going the right way. Thank you for working on this!

See my inline comments and let me know if you have any question.

Also note that in the CLI we are using the `langruntime` package as well (cmd/kubeless/function/deploy and update). In that case I think we should try to obtain the ConfigMap but since it is used just to list the available runtime when running `--help` I would not throw an error (the `kubeless` namespace may be protected and be inaccessible from the client side).

paolomainardi committed:
updated documentation about rebuilding k8s manifests file (#563)

ngtuna committed:
enable pubsub-nodejs test

ngtuna committed:
Merge pull request #545 from ngtuna/topic

fix topic missing in pubsub tests

sayanh committed:
Change service port name to use prefix http to support Istio (#560)

* Change service port name to use prefix http to support Istio

* Added http-function-port in documentation

rakesh-garimella committed:
Made changes based on comments

 - The langruntimes is now part of the controller
   data structure so that it can be passed on to
   k8sutils.go

- The langruntimes reads the configmap on the
  start of the controller.

- Currently monitoring of the configmap is missing.
  So if the configmap is changed the controller needs
  to be restarted.

- The kubeless binary has been updated to read the
  configmap everytime the kubeless binary is used to
  deploy/update the function

- Updated the unit tests

sayanh commented on cmd/kubeless/function/update.go:
```diff
@@ -170,7 +174,10 @@ var updateCmd = &cobra.Command{
 func init() {
-	updateCmd.Flags().StringP("runtime", "", "", "Specify runtime. Available runtimes are: "+strings.Join(langruntime.GetRuntimes(), ", "))
+	cli := utils.GetClientOutOfCluster()
+	var lr = langruntime.New(cli, "kubeless", "kubeless-config")
```
General comment: So we have to restart controller to change a runtimes configmap as it is picked up during bootup. Is this the expected behavior?

sayanh commented on cmd/kubeless/function/update.go:
```diff
@@ -61,9 +66,9 @@ var updateCmd = &cobra.Command{
 		}
 
-		if runtime != "" && !langruntime.IsValidRuntime(runtime) {
+		if runtime != "" && !lr.IsValidRuntime(runtime) {
```
Personally, I would love `langruntime` over `lr`

sayanh commented on cmd/kubeless/function/function.go:
```diff
@@ -258,7 +258,7 @@ func getFunctionDescription(cli kubernetes.Interface, funcName, ns, handler, fil
 		Ports: []v1.ServicePort{
 			{
-				Name:     "function-port",
+				Name:     "http-function-port",
```
Looks like git manipulation is needed as `http-` is not part of this change

sayanh commented on examples/Makefile:
```diff
@@ -239,6 +239,7 @@ pubsub-python-verify:
 	$$found
 
 pubsub-python34:
+	kubeless topic create s3-python34 || true
```
Why is this needed?
